logstash-core/src/main/java/org/logstash/instrument/metrics/BaseFlowMetric.java

```diff
@@ -0,0 +1,140 @@
+     *
+     * @return a {@link Optional} that contains a {@link FlowCapture} IFF one can be created.
+     */
+    protected Optional<FlowCapture> doCapture() {
+        final Number numeratorValue = numeratorMetric.getValue();
+        if (Objects.isNull(numeratorValue)) {
+            LOGGER.trace("FlowMetric({}) numerator metric {} returned null value", name, numeratorMetric.getName());
+            return Optional.empty();
+        }
+
+        final Number denominatorValue = denominatorMetric.getValue();
+        if (Objects.isNull(denominatorValue)) {
+            LOGGER.trace("FlowMetric({}) numerator metric {} returned null value", name, denominatorMetric.getName());
+            return Optional.empty();
+        }
+
+        final FlowCapture currentCapture = new FlowCapture(nanoTimeSupplier.getAsLong(), numeratorValue, denominatorValue);
+
+        // if our lifetime baseline has not been set, set it now.
+        if (lifetimeBaseline.offer(currentCapture)) {
+            LOGGER.trace("FlowMetric({}) baseline -> {}", name, currentCapture);
+        }
+
+        return Optional.of(currentCapture);
+    }
+
+    protected void injectLifetime(final FlowCapture currentCapture, final Map<String, Double> rates) {
+        calculateRate(currentCapture, lifetimeBaseline::get).ifPresent((rate) -> rates.put(LIFETIME_KEY, rate));
+    }
+
+    /**
+     * @param current the most-recent {@link FlowCapture}
+     * @param baseline a non-null {@link FlowCapture} from which to compare.
+     * @return an {@link OptionalDouble} that will be non-empty IFF we have sufficient information
+     * to calculate a finite rate of change of the numerator relative to the denominator.
+     */
+    protected static OptionalDouble calculateRate(final FlowCapture current, final FlowCapture baseline) {
+        Objects.requireNonNull(baseline, "baseline");
+        if (baseline == current) { return OptionalDouble.empty(); }
+
+        final BigDecimal deltaNumerator = current.numerator().subtract(baseline.numerator());
+        final BigDecimal deltaDenominator = current.denominator().subtract(baseline.denominator());
+
+        if (deltaDenominator.signum() == 0) {
+            return OptionalDouble.empty();
+        }
+
+        final BigDecimal rate = deltaNumerator.divide(deltaDenominator, LIMITED_PRECISION);
+
+        return OptionalDouble.of(rate.doubleValue());
+    }
+
+    /**
+     * @param current the most-recent {@link FlowCapture}
+     * @param possibleBaseline a {@link Supplier}{@code <FlowCapture>} that may return null
+     * @return an {@link OptionalDouble} that will be non-empty IFF we have sufficient information
+     * to calculate a finite rate of change of the numerator relative to the denominator.
+     */
+    protected static OptionalDouble calculateRate(final FlowCapture current, final Supplier<FlowCapture> possibleBaseline) {
+        return Optional.ofNullable(possibleBaseline.get())
```

Comment:
We are using `Supplier` (mostly expect to call method) like `Optional` here. If so, can we replace?